### PR TITLE
test(data): add tests for dataviewer API client and shared utility libraries (#215, #216)

### LINKED ISSUE #215: test(dataviewer): add tests for API client and data layer
## Problem Statement

The API client layer (`src/api/` or `src/services/`) and React Query hooks manage all server communication — REST endpoints, WebSocket connections, and data caching. These have no dedicated tests. Untested API code risks silent regressions in request formatting, error handling, and cache invalidation logic.

**Parent tracking issue:** #210
**Depends on:** #212 (Happy DOM + coverage config)

## Scope

Files to test:
- API client modules (fetch wrappers, endpoint functions)
- React Query hook wrappers (`useQuery`/`useMutation` configurations)
- WebSocket client utilities
- Request/response type transformations
- Error handling and retry logic

## Acceptance Criteria

- [ ] Every API client module has a corresponding `.test.ts` file
- [ ] React Query hooks tested with `@tanstack/react-query` test utilities (QueryClient wrapper)
- [ ] Tests mock `fetch`/WebSocket at the network boundary — no real HTTP calls
- [ ] Tests cover: success responses, error responses (4xx, 5xx), network failures, request parameter formatting
- [ ] All new tests pass with `npm run test`
- [ ] Coverage for API/service layer reaches 80%+

## Test Pattern

```ts
import { QueryClient, QueryClientProvider } from &#39;@tanstack/react-query&#39;
import { renderHook, waitFor } from &#39;@testing-library/react&#39;

const createWrapper = () =&gt; {
  const queryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  })
  return ({ children }: { children: React.ReactNode }) =&gt; (
    {children}
  )
}

describe(&#39;useEpisodes&#39;, () =&gt; {
  it(&#39;should fetch episodes&#39;, async () =&gt; {
    vi.spyOn(globalThis, &#39;fetch&#39;).mockResolvedValueOnce(
      new Response(JSON.stringify([{ id: &#39;1&#39; }]), { status: 200 })
    )

    const { result } = renderHook(() =&gt; useEpisodes(), {
      wrapper: createWrapper(),
    })

    await waitFor(() =&gt; expect(result.current.isSuccess).toBe(true))
    expect(result.current.data).toEqual([{ id: &#39;1&#39; }])
  })

  it(&#39;should handle fetch errors&#39;, async () =&gt; {
    vi.spyOn(globalThis, &#39;fetch&#39;).mockRejectedValueOnce(new Error(&#39;Network error&#39;))

    const { result } = renderHook(() =&gt; useEpisodes(), {
      wrapper: createWrapper(),
    })

    await waitFor(() =&gt; expect(result.current.isError).toBe(true))
  })
})
```

### Phase 1: Research — `/task-research`

```text
/task-research topic=&#34;API client and data layer in src/dataviewer/frontend/src/ — find all API modules, React Query hooks, WebSocket utilities, their dependencies, and existing test patterns for network mocking&#34;
```

### Phase 2: Plan — `/task-plan`

```text
/task-plan research=&#34;{research-output}/api-client-tests.md&#34;
```

### Phase 3: Implement — `/task-implement`

```text
/task-implement plan=&#34;{plan-output}/api-client-tests-plan.instructions.md&#34;
```

### Phase 4: Review — `/task-review`

```text
/task-review plan=&#34;{plan-output}/api-client-tests-plan.instructions.md&#34;
```

**Review focus:**
- Network calls mocked at `fetch`/WebSocket level (not implementation details)
- QueryClient created fresh per test (no shared cache)
- Error scenarios tested (4xx, 5xx, network failure)

### LINKED ISSUE #216: test(dataviewer): add tests for shared utility libraries
## Problem Statement

Shared utility libraries under `src/lib/` and `src/utils/` contain pure functions for data transformation, formatting, validation, and calculation. These are the easiest files to unit-test (no React rendering, no mocking) yet most lack test coverage.

**Parent tracking issue:** #210
**Depends on:** #212 (Happy DOM + coverage config)

## Scope

Directories to test:
- `src/lib/` — shared library functions (API helpers, data transforms, formatters)
- `src/utils/` — utility functions (math, string manipulation, validators)
- `src/constants/` — exported constants and configuration (test correct values/types)

## Acceptance Criteria

- [ ] Every `.ts` file in `src/lib/`, `src/utils/`, and `src/constants/` has a corresponding `.test.ts` file
- [ ] Tests cover: expected inputs, edge cases (empty strings, zero, null, undefined), boundary values, error conditions
- [ ] Tests are pure unit tests — no React rendering, no DOM, no mocking (unless functions have side effects)
- [ ] All new tests pass with `npm run test`
- [ ] Coverage for `src/lib/`, `src/utils/`, `src/constants/` reaches 80%+

## Test Pattern

```ts
import { formatTimestamp, clampValue, parseEpisodeId } from &#39;@/lib/formatters&#39;

describe(&#39;formatTimestamp&#39;, () =&gt; {
  it(&#39;should format seconds to mm:ss&#39;, () =&gt; {
    expect(formatTimestamp(65)).toBe(&#39;01:05&#39;)
  })

  it(&#39;should handle zero&#39;, () =&gt; {
    expect(formatTimestamp(0)).toBe(&#39;00:00&#39;)
  })

  it(&#39;should handle negative values&#39;, () =&gt; {
    expect(formatTimestamp(-1)).toBe(&#39;00:00&#39;)
  })
})
```

### Phase 1: Research — `/task-research`

```text
/task-research topic=&#34;Shared utility libraries in src/dataviewer/frontend/src/lib/ and src/dataviewer/frontend/src/utils/ — list all exported functions, their signatures, and identify pure vs side-effect functions&#34;
```

### Phase 2: Plan — `/task-plan`

```text
/task-plan research=&#34;{research-output}/utility-library-tests.md&#34;
```

### Phase 3: Implement — `/task-implement`

```text
/task-implement plan=&#34;{plan-output}/utility-library-tests-plan.instructions.md&#34;
```

### Phase 4: Review — `/task-review`

```text
/task-review plan=&#34;{plan-output}/utility-library-tests-plan.instructions.md&#34;
```

**Review focus:**
- Tests are pure (no mocking framework imports)
- Edge cases covered for each function
- Test names clearly describe behavior

### PR DESCRIPTION
**Pull request by WilliamBerryiii**
# Pull Request

## Description

Adds Vitest test coverage for two complementary slices of the dataviewer frontend:

- **#215 — API client modules:** new test suites for `api/ai-analysis`, `api/detection`, and `api/export` covering positive, error, and edge cases. Adds happy-dom env, `fake-indexeddb`, jest-dom matchers, shared test setup, and per-file coverage thresholds in `vitest.config.ts`.
- **#216 — Shared utility libraries and hooks:** new Vitest suites for `lib/` utilities (`auth-config`, `edit-draft-storage`, `joint-significance` including fast-check property tests, `offline-storage`, `query-client`, `sync-queue`) and React Testing Library suites for 13 hooks (`use-ai-analysis`, `use-annotation-workflow`, `use-annotations`, `use-batch-selection`, `use-dashboard`, `use-datasets`, `use-episodes`, `use-export`, `use-keyboard-shortcuts`, `use-labels`, `use-object-detection`, `use-offline-annotations`, `use-toast`). Adds a shared `render-hook-with-query` helper using a TanStack Query test client with infinite cache and disabled retries.

Net change: 29 files, +4,383 lines, all under `data-management/viewer/frontend/` plus a one-line `.cspell.json` dictionary update.

Closes #215
Closes #216

## Type of Change

- [ ] 🐛 Bug fix (non-breaking change fixing an issue)
- [ ] ✨ New feature (non-breaking change adding functionality)
- [ ] 💥 Breaking change (fix or feature causing existing functionality to change)
- [ ] 📚 Documentation update
- [ ] 🏗️ Infrastructure change (Terraform/IaC)
- [ ] ♻️ Refactoring (no functional changes)

Test-only change — no production code, infrastructure, or documentation modified. Closest fit not present in template.

## Component(s) Affected

- [ ] `infrastructure/terraform/prerequisites/` - Azure subscription setup
- [ ] `infrastructure/terraform/` - Terraform infrastructure
- [ ] `infrastructure/setup/` - OSMO control plane / Helm
- [ ] `workflows/` - Training and evaluation workflows
- [ ] `training/` - Training pipelines and scripts
- [ ] `docs/` - Documentation

Affects `data-management/viewer/frontend/` (not enumerated in template).

## Testing Performed

- [ ] Terraform `plan` reviewed (no unexpected changes)
- [ ] Terraform `apply` tested in dev environment
- [ ] Training scripts tested locally with Isaac Sim
- [ ] OSMO workflow submitted successfully
- [ ] Smoke tests passed (`smoke_test_azure.py`)

Frontend test suite verification (this branch):

- `npm run test` — 152 vitest tests pass
- Coverage thresholds met for the new surfaces:
  - `hooks/`: 91.72% stmts / 84.36% branches / 91.87% funcs / 93.19% lines
  - `lib/`:   96.03% stmts / 89.47% branches / 97.79% funcs / 96.21% lines
  - `stores/`: 95.47% stmts / 84.95% branches / 99.04% funcs / 97.73% lines
- `npm run lint:md` — PASS
- `npm run spell-check` — 599 files / 0 issues

## Documentation Impact

- [x] No documentation changes needed
- [ ] Documentation updated in this PR
- [ ] Documentation issue filed

## Bug Fix Checklist

*Not applicable — this is a test-only PR.*

## Checklist

- [x] My code follows the [project conventions](copilot-instructions.md)
- [x] Commit messages follow [conventional commit format](instructions/commit-message.instructions.md)
- [x] I have performed a self-review
- [x] Documentation impact assessed above
- [x] No new linting warnings introduced

🧪 - Generated by Copilot

### CODE REVIEW

**Review by katriendg — APPROVED** [review]
Approving, I believe we can merge this one and then have some additional changes with 606 potentially: 

### 1. ⚠️ Duplicate Test Helpers — Consolidate `render-hook-with-query.tsx` into `test-utils.tsx`

* Files: `src/test/render-hook-with-query.tsx`, `src/test/test-utils.tsx`
* Severity: Low (code smell / maintainability)

**Problem:** Two files export identically-named `createTestQueryClient` and `renderHookWithQuery` functions with different semantics:

| File | `gcTime` | `staleTime` | Signature style |
|------|----------|-------------|-----------------|
| `test-utils.tsx` | `0` (isolated) | default | positional args |
| `render-hook-with-query.tsx` | `Infinity` (persistent) | `Infinity` | options-bag |

Seven hook tests import from `render-hook-with-query.tsx`; two tests import from `test-utils.tsx`.

**Suggested Resolution:**

1. Delete `render-hook-with-query.tsx`.
2. Add a `createPersistentQueryClient()` variant to `test-utils.tsx` with `gcTime: Infinity, staleTime: Infinity`.
3. Update the `renderHookWithQuery` export in `test-utils.tsx` to accept an optional `queryClient` in the options bag (matching the deleted file's API).
4. Update the 7 importing hook tests to use `@/test/test-utils` instead of `@/test/render-hook-with-query`.

### 2. 💡 React `act()` Warnings in Hook Tests

* Affected tests: Multiple hook test files emit "not wrapped in act(...)" console warnings.
* Severity: Low (cosmetic; all tests pass)

**Problem:** Some state updates triggered by async query resolution are not wrapped in `act()`. Currently benign but may become errors in future React versions.

**Suggested Resolution:** Wrap `waitFor` or explicit `act()` around state-triggering operations in hook tests that use `renderHookWithQuery`. Low priority — address opportunistically when touching these files.

### 3. 💡 No Per-File Coverage Thresholds for Hooks

* File: `vitest.config.ts`
* Severity: Informational

**Problem:** The PR adds 80% per-file coverage thresholds for `src/api/*.ts` and `src/lib/*.ts` modules but does not add equivalent thresholds for the 13 newly-tested hooks in `src/hooks/`.

**Suggested Resolution:** Add per-file thresholds for hooks that have achieved stable coverage (e.g., `use-keyboard-shortcuts`, `use-annotations`, `use-dashboard`). Can be deferred until hook coverage stabilizes.
